cmd/opm/serve/serve.go

```diff
@@ -103,9 +122,70 @@ func (s *serve) run(ctx context.Context) error {
 	health.RegisterHealthServer(grpcServer, server.NewHealthServer())
 	reflection.Register(grpcServer)
 	s.logger.Info("serving registry")
+	p.stopStartupCpuProfiling()
 	return graceful.Shutdown(s.logger, func() error {
 		return grpcServer.Serve(lis)
 	}, func() {
 		grpcServer.GracefulStop()
 	})
+
+}
+
+type profilerInterface struct {
+	addr string
+	cpu  bytes.Buffer
+
+	logger *logrus.Entry
+}
+
+func newProfilerInterface(a string, log *logrus.Entry) *profilerInterface {
+	return &profilerInterface{
+		addr:   a,
+		logger: log.WithFields(logrus.Fields{"address": a + ":" + defaultProfPort}),
+		cpu:    bytes.Buffer{},
+	}
+}
+
+func (p *profilerInterface) isEnabled() bool {
+	return p.addr != ""
+}
+
+func (p *profilerInterface) startDefaultEndpoint() {
+	// pprof tool HTTP endpoint
+	if !p.isEnabled() {
+		return
+	}
+
+	// only logs fatal error on listen failure, otherwise swallowed on nominal closure
+	// goroutine exits with main
+	go func() {
+		p.logger.Info("starting default pprof endpoint")
+		p.logger.Fatal(http.ListenAndServe(p.addr+":"+defaultProfPort, nil))
+	}()
+}
+
+func (p *profilerInterface) startCustomEndpoint() error {
+	if !p.isEnabled() {
+		return nil // short-circuit if not enabled
+	}
+
+	http.HandleFunc(defaultCpuStartupPath, p.httpHandler)
+	p.logger.Infof("start caching startup cpu profile data at %q", defaultCpuStartupPath)
+	if err := pprof.StartCPUProfile(&p.cpu); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *profilerInterface) stopStartupCpuProfiling() {
+	if !p.isEnabled() {
+		return
+	}
+	pprof.StopCPUProfile()
+	p.logger.Info("stop caching startup cpu profile data")
+}
+
+func (p *profilerInterface) httpHandler(w http.ResponseWriter, r *http.Request) {
+	w.Write(p.cpu.Bytes())
```

Comment:
Is it possible to query this endpoint before the profile data has been written?

If so, should we return a different http error code (maybe 503 Service Unavailable)?

Reply:
It is *technically* possible to query before any data has been written, but I felt that:

- timing is fast enough that one would have to specifically target that interval (at least for this freshman effort)
- the existing error message was sufficient: `Fetching profile over HTTP from http://localhost:60066/debug/pprof/startup/cpu http://localhost:60066/debug/pprof/startup/cpu: parsing profile: empty input file`

I understand this as the use-case.  If you think that we need to support a general access attempt instead of the specific tool (e.g. curl), then there would be a benefit to providing a different status before the data is available/complete.

Reply:
Ideally, we'd have a `p.isProfileReady` (protected by a mutex) such that we'd do:

```go
p.mutex.RLock()
isProfileReady := p.isProfileReady
p.mutex.RUnlock()
if !isProfileReady  {
	http.Error(w, errors.New("profile is not yet ready"), http.StatusServiceUnavailable)
	return
}
w.Write(p.cpu.Bytes())
```

Again, this is more of a "this is how a hardened server should work", but in my mind this is low-effort enough that we don't have to cut this corner.

Reply:
Done